include/core/CDataFrame.h

```diff
@@ -0,0 +1,406 @@
+                                 EReadWriteToStorage ayncWriteToStore,
+                                 TWriteSliceToStoreFunc writeSliceToStore);
+        ~CDataFrameRowSliceWriter();
+
+        //! Write a single row using the callback \p writeRow.
+        void operator()(const TWriteFunc& writeRow);
+
+        //! Finish writing the rows and return the number of rows written and
+        //! the slices.
+        TSizeDataFrameRowSlicePtrVecPr finishWritingRows();
+
+    private:
+        //! This is called to flush the queue to write to the store.
+        void finishAsyncWriteToStore();
+
+    private:
+        std::size_t m_NumberRows;
+        std::size_t m_RowCapacity;
+        std::size_t m_SliceCapacityInRows;
+        EReadWriteToStorage m_AsyncWriteToStore;
+        TWriteSliceToStoreFunc m_WriteSliceToStore;
+
+        TFloatVec m_SliceBeingWritten;
+
+        //! This is true while the rows are still being added.
+        bool m_Writing = true;
+        //! A queue of finished slices shared with the thread that writes
+        //! them to storage if writes to storage are asynchronous.
+        TSizeFloatVecPrQueue m_SlicesToAsyncWriteToStore;
+        //! The result of the asynchronous work to write slices to storage
+        //! if there is any.
+        std::future<TRowSlicePtrVec> m_AsyncWriteToStoreResult;
+        //! The synchronously written stored slices.
+        TRowSlicePtrVec m_SyncWrittenSlices;
+    };
+    using TRowSliceWriterPtr = std::unique_ptr<CDataFrameRowSliceWriter>;
+
+private:
+    //! Get the number of threads and the stride (in the data frame) for the
+    //! target number of threads \p target.
+    //!
+    //! \note This is always less than or equal to \p target.
+    TSizeSizePr numberOfThreadsAndStride(std::size_t target) const;
+
+private:
+    //! The number of rows in the data frame.
+    std::size_t m_NumberRows = 0;
+    //! The number of columns in the data frame.
+    std::size_t m_NumberColumns;
+    //! The number of columns a row could contain. This is greater than or
+    //! equal to m_NumberColumns.
+    std::size_t m_RowCapacity;
+    //! The capacity of a slice of the data frame as a number of rows.
+    std::size_t m_SliceCapacityInRows;
+
+    //! If true read and write asynchronously to storage.
+    EReadWriteToStorage m_AsyncReadAndWriteToStore;
+    //! The callback to write a slice to storage.
+    TWriteSliceToStoreFunc m_WriteSliceToStore;
+
+    //! The stored slices.
+    TRowSlicePtrVec m_Slices;
+
+    //! The slice writer which is currently active.
+    TRowSliceWriterPtr m_Writer;
+};
+
+//! Make a data frame which uses main memory storage for its slices.
+//!
+//! \param[in] numberColumns The number of columns in the data frame created.
+//! \param[in] sliceCapacity If none null this overrides the default slice
+//! capacity in rows.
+//! \param[in] readWriteToStoreAsync Controls whether reads and writes from
```

Comment:
maybe `readWriteToStoreSyncStrategy` or `readWriteToStoreSyncOptions`? - right now it sounds like a bool.